### ISSUE TITLE
CA1508 false positive in while condition with String.IsNullOrEmpty() and initially empty string

### ISSUE THREAD

**Issue by JeremyCaney** [dotnet/roslyn-analyzers#4546]
### Analyzer

**Diagnostic ID**: [CA1508](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1508): Avoid dead conditional code

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 5.0.100](https://dotnet.microsoft.com/download/dotnet/5.0)

### Describe the bug

When evaluating a `while()` loop with a `String.IsNullOrEmpty()` condition, the expression is evaluated as "always 'false'" if the variable starts out empty, even though the initial state will be `true`. 

### Steps To Reproduce

The following is a minimum reproducible example that demonstrates this behavior:
```cs
string? myString = String.Empty; 
while (String.IsNullOrEmpty(myString)) 
{
  myString = "Hello, World";
}
```

Similarly, if I flip the logic:

```cs
string? myString = "Hello, World";
while (!String.IsNullOrEmpty(myString)) 
{
  myString = String.Empty;
}
```

> _Note:_ It doesn't matter if the initial value is `""` or `String.Empty`; I'm just using the static field because it's more obvious.

### Expected behavior

Since the initial value of `myString` is empty, `String.IsNullOrEmpty()` will return `true` on the first evaluation; it is then set to a non-empty value within the `while()` loop, and so returns `false` on the second evaluation. Given this, I would not expect CA1508 to issue a warning. 

### Actual behavior

In this scenario, CA1508 issues the following warning:

> 'String.IsNullOrEmpty(myString)' is always 'false'. Remove or refactor the condition(s) to avoid dead code.

## Additional context

Curiously, this works exactly as expected if the initial value is `null` instead of empty:

```cs
string? myString = null;
while (String.IsNullOrEmpty(myString)) 
{
  myString = "Hello, World";
}
```

And, again, if I flip the logic:

```cs
string? myString = "Hello, World";
while (!String.IsNullOrEmpty(myString)) 
{
  myString = null;
}
```

As such, this appears to be specific to how `String.IsNullOrEmpty()` is evaluating an initially empty string when in a `while()` loop.

**Comment by JeremyCaney** [comment]
Aside, there is a minor bug with the label of this warning which might confuse this issue. In the following similar code, we _expect_ the CA1508 to be thrown:

```cs
string myString = String.Empty;
if (String.IsNullOrEmpty(myString) 
{
}
```

And, indeed, it is. But the warning incorrectly states:

> 'String.IsNullOrEmpty(myString)' is always 'false'. Remove or refactor the condition(s) to avoid dead code.

Obviously, it should say "is always 'true'". That's a separate issue—and one I'm happy to create a bug report for!—but I wanted to point it out here since it might otherwise obscure the more important issue with the `while()` loop.

**Comment by JeremyCaney** [comment]
Also, this is probably obvious, but as a downstream consequence of this, any other `&&` conditions that are paired with it will also throw a CA1508: 
```cs
string? myString = String.Empty;
object? myObject = new();
while (String.IsNullOrEmpty(myString) && myObject is not null) 
{
  myObject = null;
  myString = "Hello, World";
}
```
In this case, the `myObject is null` would normally be fine if it weren't paired with the `String.IsNullOrEmpty(myString)` condition. Since the analyzer is seeing `myString` as always being false, however, it's evaluating it the same as:
```cs
object? myObject = new();
while (false && myObject is not null) 
{
   myObject = null;
}
```

In this case, the analyzer will throw a CA1508 warning on `myObject is not null`, since the `false` invalidates the subsequent condition.

**Comment by Evangelink** [comment]
As far as I understand the current DFA doesn't really handle loops.

Tagging @mavasani

**Comment by mavasani** [comment]
Verified that all repro cases are fixed in master. We have a released `Microsoft.CodeAnalysis.NetAnalyzers` [NuGet package 5.0.1](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers/5.0.1) which has the fix. However, the fixed analyzer has not been inserted into .NET SDK, hence you will continue seeing this issue for some more time if you are consuming the analyzers directly from the SDK (unless you want to try out [.NET 6 SDK builds](https://github.com/dotnet/installer#installers-and-binaries)).

Closing the issue.

**Comment by Evangelink** [comment]
@mavasani I am reopening because I think you did the same error I did when reading this. It's more a false-negative than false-positive, no diagnostic is reported when it should be reported.

**Comment by mavasani** [comment]
@Evangelink The original issue comment says the Expected behavior is no warning, but actual behavior is warning. I verified that it is indeed not generating any warning. If there are any false negative cases also mentioned in intermediate comments, please open a separate issue with an appropriate title and only the relevant repro case.

**Comment by JeremyCaney** [comment]
@mavasani: I've installed the 5.0.1 version of the NuGet package and confirmed that the CA1508 false positive is resolved. I apologize for missing the resolution; in the future, I'll be sure to test against the latest NuGet package, instead of relying on the lagging SDK version. Thank you for all of your work on these!
